### ISSUE TITLE
gpu-executionにDockerへGPUをCDIで渡す契約を記載する

### ISSUE THREAD

**Issue by iwashita-nozomu** [iwashita-nozomu/agent-canon#748]
## 概要

AgentCanon の `gpu-execution` skill に、Docker containerへGPUを渡すときはCDIを使う、と明記してください。

rootful/rootless、legacy/CDIを判定して分岐する新しいruntime機構は不要です。既存のGPU admissionが空きGPUを選択・lockし、そのcommandがDockerを起動する場合のdevice wiringをCDIに統一するだけです。

### DockerへのGPU device受け渡し

Docker containerをGPU childとして起動する場合は、`--gpus` やNVIDIA legacy hookではなくCDI deviceを渡します。

```bash
docker run --rm \
  --device nvidia.com/gpu=all \
  -e CUDA_VISIBLE_DEVICES \
  -e NVIDIA_VISIBLE_DEVICES \
  -e JAX_PLATFORMS \
  -e XLA_PYTHON_CLIENT_PREALLOCATE \
  -e XLA_PYTHON_CLIENT_ALLOCATOR \
  -e XLA_PYTHON_CLIENT_USE_CUDA_HOST_ALLOCATOR \
  <canonical-image> <argv...>
```

`run_gpu_command.py` がplan freeze後にDocker CLI childへ設定した環境は、Dockerの `-e NAME` で値を変更せずcontainer childへ転送します。

- AgentCanonが選択するGPU: full UUID
- `CUDA_VISIBLE_DEVICES`: selected full UUID
- `NVIDIA_VISIBLE_DEVICES`: 同じfull UUID
- Dockerへ渡すdevice: CDI
- JAX import: container内のfresh childで初めて行う
- CPU fallback: 認めない

CDI registryが個別GPU handleを提供する環境では、そのCDI handleを渡せます。WSLで `nvidia.com/gpu=all` だけが生成される場合は `all` をcontainerへinjectし、計算対象は既存のfull UUID environmentで限定します。

## cgroupについて書く内容

cgroup全体を無効化する指示は不要です。

今回失敗したのは、rootless WSLで `docker run --gpus ...` がNVIDIA legacy hookを選び、次のdevice-cgroup更新を試みたためです。

```text
open /sys/fs/cgroup/devices/user.slice/devices.allow: permission denied
```

CDIでGPUを渡せば、このlegacy device-cgroup操作をGPU wiringの前提にする必要がありません。

`no-cgroups=true` を設定する手順や、rootful/rootlessを判定する分岐はskillへ追加しません。

## 確認済みの実行結果

環境:

- WSL 2 runtime 2.4.11
- Ubuntu 24.04.3 LTS
- kernel `5.15.167.4-microsoft-standard-WSL2`
- rootless Docker 29.7.2
- NVIDIA Container Toolkit 1.18.1
- NVIDIA RTX 4000 Ada Generation x3

CDI spec:

```bash
mkdir -p ~/.config/cdi
nvidia-ctk cdi generate --output ~/.config/cdi/nvidia.yaml
```

実行:

```bash
docker run --rm \
  --device nvidia.com/gpu=all \
  python:3.11-slim-bookworm \
  nvidia-smi -L
```

3 GPUの認識に成功しました。selected full UUIDを環境変数で転送したcontainer内JAX 0.10.1でも次を確認済みです。

```text
default_backend=gpu
devices=[CudaDevice(id=0)]
```

## 二重mountとの区別

Docker contextの次のendpoint表記は、二重mountではありません。

```text
unix:///run/user/1000//docker.sock
```

shellとrootless Docker daemonのmount namespaceを確認し、`/run/user/1000` は単一のtmpfs mountでした。余分な `/` はpath separatorの表記だけです。

## Scope

主対象:

- `agents/skills/gpu-execution.md`

整合が必要な既存記述:

- `agents/skills/environment-maintenance.md` の `docker run --gpus ...` 表記を、`gpu-execution` のCDI契約参照へ合わせる

これは新しいprovider、scheduler、admission state machine、Docker runtime detectorを追加するIssueではありません。

## 完了条件

- `gpu-execution` skillが「DockerへGPUを渡す場合はCDI」と明記する。
- `--device nvidia.com/gpu=...` とexact environment forwardingの実行例がある。
- `--gpus` / legacy hookをcanonical Docker GPU commandにしない。
- cgroup全体の無効化や `no-cgroups=true` を要求しない。
- rootful/rootlessのroute-selection実装を追加しない。
- selected GPUはfull UUIDのままで、integer indexへ変換しない。
- container内JAX GPU smokeをvalidation例として記載する。
- `environment-maintenance` の既存 `--gpus` 文言と矛盾しない。

## 関連

- #737: provider-independent direct GPU admission
- #746: `nvidia-smi` XML DOCTYPE parser bug
- jax_utils #488: canonical Docker imageとdriver compatibility

**Comment by iwashita-nozomu** [comment]
## 対応結果

PR #780 で修正しました。

- Branch: `fix/747-748-update-entrypoint-cdi`
- Base: `main@e075558dff342c64339ab7c692b10c183d97f91c`
- Head: `37a9c298062d2d27c9d18d63a63509d3c37c010e`
- PR: #780

## 設計方針

このIssueは新しいprovider、scheduler、Docker runtime detectorを作るものではなく、既存GPU admissionが選択・lockしたphysical GPUをDocker childへ渡す境界だけを一意にする変更として実装しました。

canonical Docker GPU routeは次の形です。

```text
docker run --rm \
  --device nvidia.com/gpu=<qualified-device> \
  -e CUDA_VISIBLE_DEVICES \
  -e NVIDIA_VISIBLE_DEVICES \
  -e JAX_PLATFORMS \
  -e XLA_PYTHON_CLIENT_PREALLOCATE \
  -e XLA_PYTHON_CLIENT_ALLOCATOR \
  -e XLA_PYTHON_CLIENT_USE_CUDA_HOST_ALLOCATOR \
  <canonical-image> <argv...>
```

## 追加した契約

`agents/skills/gpu-execution.md` にDocker GPU wiring節を追加し、以下を固定しました。

1. device injectionはCDIの `--device nvidia.com/gpu=<qualified-device>` を使う。
2. admission identityはselected full GPU UUIDのまま保持し、integer indexやUUID prefixへ変換しない。
3. CDI registryが個別GPU qualified deviceを公開する場合は、そのdeviceを優先する。
4. WSL等で `nvidia.com/gpu=all` しか公開されない場合はdevice injectionに`all`を使い、compute visibilityはfull UUIDの `CUDA_VISIBLE_DEVICES` と `NVIDIA_VISIBLE_DEVICES` で限定する。
5. admissionが固定した6変数は値を書き換えず、同じDocker argvで `-e NAME` として全て転送する。
6. host側や既存container stateでJAXを先にimportせず、新規container child内でfresh importする。
7. `jax.default_backend() != "gpu"` はCPU fallbackなのでGPU validation passにしない。
8. CDI unavailable/device injection failureはtyped blockerとしてcloseoutへ残す。
9. `no-cgroups=true`、cgroup全体無効化、rootful/rootless自動判定、legacy hook/CDI選択detectorは追加しない。
10. Docker socket URIの重複slashと二重mount evidenceを混同しない。

`agents/skills/environment-maintenance.md` の既存 `docker run --gpus ...` 表記も削除し、GPU image validationは`gpu-execution`が所有するCDI admission、full UUID visibility、exact environment forwarding、container内fresh JAX smokeへ委譲しました。これによりenvironment ownerとGPU execution ownerの指示が矛盾しません。

## 変更ファイル

- `agents/skills/gpu-execution.md`
- `agents/skills/environment-maintenance.md`
- `tests/agent_tools/test_gpu_execution_cdi_contract.py`

PRには#747との共通変更を含め、全体で6ファイルです。このIssueのowner差分は上記3ファイルです。

## 検証

focused documentation-contract regression 4件を追加・実行し、全件成功しました。

- CDI device injectionと6個のexact environment名がcanonical skillに存在する
- WSL `nvidia.com/gpu=all` fallbackでもcompute visibilityがfull UUIDに固定される
- `docker run --gpus`、`no-cgroups=true`、`nvidia-container-runtime` をcanonical記述へ戻さない
- `environment-maintenance` がCDI commandとcontainer内fresh JAX smokeを`gpu-execution`へ委譲する

併せて次を確認しました。

- `python3 -m py_compile` 成功
- `git diff --check` 成功
- 既存Issueに記録済みのWSL/rootless Docker/CDI/JAX実機結果と矛盾しない契約に限定
- 新しいruntime分岐・provider state・cgroup変更コードを追加していない
- PRは最新mainの直上1commit、behind 0、意図した6ファイルのみ
- GitHub Actions: `Issue Mirror`、`Agent Runtime Dashboard`、`Entrypoint Owner Map`、`AgentCanon Static Gates` がすべてsuccess
- PRはnon-draft、mergeable、未解決review/threadなし

この変更は文書ownerと回帰契約の修正であり、新しいGPU runtime実装ではありません。Issue本文の実機evidenceを前提に、repository側では将来のlegacy route再混入をstatic regressionで防止しています。この状態でレビュー・マージ可能です。

**Comment by iwashita-nozomu** [comment]
## 完了報告 — PR #780 merged

Issue #748 の Docker CDI GPU 契約は PR #780 で完了し、2026-08-19 04:37:36 UTC に `main` へ squash merge されました。

- PR: #780
- 作業 branch: `fix/747-748-update-entrypoint-cdi`
- 検証済み最終 PR head: `43ae36ea18c91d10414b7bdf205a7aac68e7488c`
- merge 前 current main: `6c8bc4a57103a6e3b71b8603cd22a411a569741e`
- final main / squash SHA: `0ea5bb6d5d0bfc2e027698612aeb6fc5a3c8b0c2`
- final tree: `7461138908201123e5cf1cc6e82740c99b9857a8`

### current-main drift と統合

PR の当初 base `e075558dff342c64339ab7c692b10c183d97f91c` から、作業中に PR #779 が merge され、`main` は `6c8bc4a57103a6e3b71b8603cd22a411a569741e` へ進みました。

#779 の routing-owner 変更8ファイルと #780 の6ファイルには変更 path の重複がなく、blob/mode を保持して既存 branch へ merge commit `43ae36ea18c91d10414b7bdf205a7aac68e7488c` として統合しました。最終比較は `ahead 6 / behind 0` で、current-main conflict や semantic drift はありませんでした。

squash merge 後の `6c8bc4a571… → 0ea5bb6d5d0…` は1 commit・6ファイルだけです。一時的な修正適用 workflow は final tree に含まれません。

### 最終契約

`agents/skills/gpu-execution.md` を Docker GPU wiring の canonical owner とし、次を明文化しました。

1. Docker container へ GPU device を渡す canonical route は CDI です。

```bash
docker run --rm \
  --device nvidia.com/gpu=<qualified-device> \
  ...
```

2. AgentCanon admission が選択する device identity は physical/MIG の **full GPU UUID** のままです。integer index や UUID prefix へ変換しません。
3. CDI registry が selected full UUID に対応する個別 qualified device を提供する場合は、その handle を `--device` へ渡します。
4. WSL 等で registry が `nvidia.com/gpu=all` だけを公開する場合は device-node injection に `all` を使いますが、compute visibility の authority は admission 済み full UUID のままです。
5. admitted child が固定した次の6変数を、値を書き換えず同一 Docker argv の `-e NAME` で全て転送します。

```text
CUDA_VISIBLE_DEVICES
NVIDIA_VISIBLE_DEVICES
JAX_PLATFORMS
XLA_PYTHON_CLIENT_PREALLOCATE
XLA_PYTHON_CLIENT_ALLOCATOR
XLA_PYTHON_CLIENT_USE_CUDA_HOST_ALLOCATOR
```

6. JAX は host や既存 container state で先に import せず、新規 container child 内で初めて import します。`jax.default_backend() == "gpu"` と device readback が GPU validation oracle で、CPU fallback は pass にしません。
7. CDI device injection failure は typed blocker として閉じます。cgroup 全体を無効化して通す route にはしません。
8. Docker socket URI の重複 slash と二重 mount を混同せず、socket mount evidence と CDI device evidence を別の事実として扱います。

`agents/skills/environment-maintenance.md` は GPU wiring の実装判断を重複所有せず、`gpu-execution.md` の CDI/full UUID/exact environment 契約へ委譲する形へそろえました。

### 復活させていない経路

最終差分を確認し、次を canonical command または実装として追加していません。

- `docker run --gpus ...`
- NVIDIA legacy runtime hook を選択する route
- `nvidia-container-runtime`
- `no-cgroups=true`
- rootful/rootless の自動 route selection
- CDI と legacy hook を切り替える第二 runtime selector
- 第二 scheduler/provider/GPU registry
- 第二 GPU admission state machine

既存 `run_gpu_command.py` の provider-independent admission、full UUID lock、plan freeze、exact environment をそのまま owner とし、この Issue では Docker child 境界の device wiring と validation contract だけを追加しています。

### 差分

main 上の最終差分は次の6ファイルです。

- `agents/skills/agent-canon-update.md`
- `agents/skills/environment-maintenance.md`
- `agents/skills/gpu-execution.md`
- `tests/agent_tools/test_gpu_execution_cdi_contract.py`
- `tests/agent_tools/test_skill_tool_commands_update_entrypoint.py`
- `tools/agent_tools/skill_tool_commands.py`

#748 の直接 owning surface は `gpu-execution.md`、委譲 consumer の `environment-maintenance.md`、CDI documentation regression の3ファイルです。GPU runtime implementation 自体は変更していません。

### validation

focused validation:

- `python3 -m py_compile tools/agent_tools/skill_tool_commands.py tests/agent_tools/test_skill_tool_commands_update_entrypoint.py tests/agent_tools/test_gpu_execution_cdi_contract.py` — pass
- `python3 -m unittest tests.agent_tools.test_skill_tool_commands_update_entrypoint tests.agent_tools.test_gpu_execution_cdi_contract -v` — pass
  - update-entrypoint: 10 tests
  - CDI/document contract: 4 tests
- `git diff --check` — pass

CDI regression 4 tests は次を固定しています。

- `--device nvidia.com/gpu=...` と6変数の exact forwarding
- WSL の `all` injection と full UUID compute visibility の分離
- owning documents に canonical `docker run --gpus`、`no-cgroups=true`、`nvidia-container-runtime` が存在しないこと
- `environment-maintenance` が `gpu-execution` へ委譲し、fresh in-container JAX verification を要求すること

最終 head `43ae36ea18c91d10414b7bdf205a7aac68e7488c` の GitHub Actions readback:

- AgentCanon Static Gates — success, run `32216145557`
  - `select-static-units`: success
  - `contracts-static`: success
  - aggregate `static-gates`: success
- Entrypoint Owner Map — success, run `32216145555`
  - entrypoint grammar: success
  - entrypoint regression tests: success
- Agent Runtime Dashboard — success, run `32216145584`
- Issue Mirror — success, run `32216145563`

review readback は submitted review なし、unresolved review thread なし、PR conversation comment なしでした。

### GPU verification boundary

今回の final readback は GitHub connector 上で行っており、新しい物理 GPU 実行は行っていません。この変更は GPU runtime 実装ではなく、既存 admission から Docker child へ渡す契約とその静的 regression を所有するためです。

一方、Issue 本文には同じ route の実機 evidence が既に記録されています。

- rootless Docker + WSL で `--device nvidia.com/gpu=all` により `nvidia-smi -L` が3 GPUを認識
- selected full UUID を environment で制限した fresh container 内 JAX で `default_backend=gpu`
- `devices=[CudaDevice(id=0)]` と exactly one visible device を確認

PR はこの実測 route を正本化し、今後の live closeout でも fresh in-container JAX GPU probe を必須 oracle としました。static test や CPU-only smoke を GPU pass とする余地は残していません。このため本 Issue に残る `need verification` はありません。

### merge evidence / status

PR #780 は expected head `43ae36ea18c91d10414b7bdf205a7aac68e7488c` を指定した CAS 付き squash merge で成功しました。GitHub readback では PR は `closed` / `merged=true`、merge SHA は `0ea5bb6d5d0bfc2e027698612aeb6fc5a3c8b0c2`、`main` HEAD も同一 SHA です。Issue は `completed` として自動 close され、`in progress` は解除済み、`ready for review` を保持しています。
